normandy/recipes/models.py

```diff
@@ -77,64 +80,84 @@ class Recipe(models.Model):
-
-        if field_vals == []:
-            return True
-
-        for val in field_vals:
-            if val.matches(client_val):
-                return True
-
-        field_vals_str = ', '.join(str(o) for o in field_vals)
-        self.log_rejection('client {name} ({client_val}) does not match recipe '
-                           '(choices are {field_vals})'
-                           .format(name=name, client_val=client_val, field_vals=field_vals_str))
-        return False
+    @classmethod
+    def register_matcher(cls, func):
+        cls._registered_matchers.append(func)
 
     def matches(self, client):
         """
         Return whether this Recipe should be sent to the given client.
         """
-        # This should be ordered roughly by performance cost
-        if not self.enabled:
-            self.log_rejection('not enabled')
-            return False
+        for matcher in self._registered_matchers:
+            if not matcher(self, client):
+                return False
 
-        if self.start_time and self.start_time > client.request_time:
-            self.log_rejection('start time not met ({})'.format(self.start_time))
-            return False
+        return True
 
-        if self.end_time and self.end_time < client.request_time:
-            self.log_rejection('end time already passed ({})'.format(self.end_time))
-            return False
+    def __repr__(self):
+        return '<Recipe "{name}">'.format(name=self.name)
 
-        if self.sample_rate:
-            inputs = [self.pk, client.user_id]
-            if not utils.deterministic_sample(self.sample_rate / 100.0, inputs):
-                self.log_rejection('did not match sample')
-                return False
+    def __str__(self):
+        return self.name
 
-        if not self.check_many('locale', self.locales, client.locale):
-            return False
 
-        if not self.check_many('country', self.countries, client.country):
-            return False
+@Recipe.register_matcher
+def match_enabled(recipe, client):
+    return recipe.enabled
+
 
-        if not self.check_many('channel', self.release_channels, client.release_channel):
+@Recipe.register_matcher
+def match_times(recipe, client):
+    if recipe.start_time and recipe.start_time > client.request_time:
+        return False
+
+    if recipe.end_time and recipe.end_time < client.request_time:
+        return False
+
+    return True
+
+
+@Recipe.register_matcher
+def match_sample_rate(recipe, client):
+    if recipe.sample_rate:
+        inputs = [recipe.pk, client.user_id]
+        if not utils.deterministic_sample(recipe.sample_rate / 100.0, inputs):
             return False
 
-        return True
+    return True
 
-    def __repr__(self):
-        return '<Recipe "{name}">'.format(name=self.name)
 
-    def __str__(self):
-        return self.name
+def multivalue_matcher(field_getter):
+    """Make a matcher that accepts the client if any of the recipe's values accept the client."""
```

Comment:
I think this docstring should document that we expect a `matches` function on the values we're comparing and what arguments it should take.